src/main/java/teammates/common/util/LogMessageGenerator.java

```diff
@@ -0,0 +1,252 @@
+import teammates.common.util.ActivityLogEntry.Builder;
+
+/**
+ * Factory to generate log message.
+ */
+public class LogMessageGenerator {
+    public static final Pattern PATTERN_ACTION_NAME = Pattern.compile("^/\\S+?/(?<actionName>[^\\s\\?]*)");
+    public static final String PATTERN_ACTION_NAME_GROUP = "actionName";
+
+    /**
+     * Generates log message for all *Action page.
+     *
+     * @param url URL of the action
+     * @param params parameterMap of the request
+     * @param currUser login information generated by {@link teammates.logic.api.GateKeeper}
+     * @param userAccount authentication user account generated by action
+     * @param unregisteredStudent authentication unregisteredStudent attributes generated by action
+     * @param logMessage log message to show to admin
+     * @return log message in the form specified in {@link ActivityLogEntry}
+     */
+    public String generatePageActionLogMessage(String url, Map<String, String[]> params, UserType currUser,
+            AccountAttributes userAccount, StudentAttributes unregisteredStudent, String logMessage) {
+        Builder builder = generateBasicActivityLogEntry(url, params, currUser);
+
+        boolean isUnregisteredStudent = unregisteredStudent != null;
+        boolean isAccountWithGoogleId = userAccount != null && userAccount.googleId != null;
+        if (isUnregisteredStudent) {
+            updateInfoForUnregisteredStudent(builder, unregisteredStudent);
+        } else if (isAccountWithGoogleId) {
+            updateInfoForNormalUser(builder, currUser, userAccount);
+        }
+
+        builder.withLogMessage(logMessage);
+        return builder.build().generateLogMessage();
+    }
+
+    private void updateInfoForUnregisteredStudent(Builder builder, StudentAttributes unregisteredStudent) {
+        String role = Const.ActivityLog.ROLE_UNREGISTERED;
+        if (unregisteredStudent.course != null && !unregisteredStudent.course.isEmpty()) {
+            role = Const.ActivityLog.ROLE_UNREGISTERED + ":" + unregisteredStudent.course;
+        }
+        builder.withUserRole(role)
+               .withUserName(unregisteredStudent.name)
+               .withUserEmail(unregisteredStudent.email);
+    }
+
+    private void updateInfoForNormalUser(Builder builder, UserType currUser, AccountAttributes userAccount) {
+        checkAndUpdateForMasqueradeMode(builder, currUser, userAccount);
+        builder.withUserGoogleId(userAccount.googleId)
+               .withUserEmail(userAccount.email)
+               .withUserName(userAccount.name);
+    }
+
+    private void checkAndUpdateForMasqueradeMode(Builder builder, UserType loggedInUser, AccountAttributes account) {
+        if (loggedInUser != null && loggedInUser.id != null && account != null) {
+            boolean isMasqueradeMode = !loggedInUser.id.equals(account.googleId);
+            builder.withMasqueradeUserRole(isMasqueradeMode);
+            // TODO discuss: should we rebuild log id?
```

Comment:
Yes, I'm ok with using different IDs for admin. However, when using the 'view recent actions by the user' feature in the admin log page,  it may be useful to show both user actions and admin actions (on behalf of that user) because both modify the same user data and both are responsible for the current state of the data.